File: py/optimize/steepest_descent_optimizer_test.py
import py.optimize.target_function as tf
import py.optimize.steepest_descent_optimizer as sdo
import numpy as np


class TestTargetFunction(tf.TargetFunction):

    # f(x1, x2) = x1^2 + x2
    def fun(self, params, args):
        x1 = args[0]
        x2 = args[1]
        res = np.add(np.multiply(params[0], np.power(x1, 2)), np.multiply(params[1], x2))
        return res


class AnotherTestTargetFunction(tf.TargetFunction):

    def fun(self, params, args):
        x = args[0]
        a = params[0]
        b = params[1]
        res = a * x + b
        return res


def test_optimizer():
    params = np.array([0.1, 0.1])
    train_input = np.array([[-1], [0], [1]])
    truth_output = np.array([[-1], [1], [3]])
    target_function = AnotherTestTargetFunction()
    optimizer = sdo.SteepestDescentOptimizer()
    optimized_params = optimizer.optimize(target_function, params, train_input, truth_output)
    print(optimized_params)

def test_calc_partial_derivative():
    params = np.array([1.0, 2.0])
    args = np.array([4.0, 3.0])
    target_function = TestTargetFunction()
    optimizer = sdo.SteepestDescentOptimizer()
    partial_derivative = optimizer.calc_partial_derivative(target_function, params, 0, args)
    print(partial_derivative)

if __name__ == "__main__":
    test_optimizer()
    test_calc_partial_derivative()

File: src/test/java/jm/app/optimize/SteepestDescentOptimizerTest.java
package jm.app.optimize;

import jm.app.algebra.Matrix;
import jm.app.optimize.SteepestDescentOptimizer;
import jm.app.optimize.TargetFunction;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

public class SteepestDescentOptimizerTest {

    private SteepestDescentOptimizer optimizer = new SteepestDescentOptimizer();

    @Test
    public void testOptimize() {
        Matrix params = new Matrix(1, 2);
        params.setValue(0, 0, 0.01);
        params.setValue(0, 1, 0.01);

        Matrix trainInput = new Matrix(3, 1);
        trainInput.setValue(0, 0, -1);
        trainInput.setValue(1, 0, 0);
        trainInput.setValue(2, 0, 1);

        Matrix truthOutput = new Matrix(3, 1);
        truthOutput.setValue(0, 0, -1);
        truthOutput.setValue(1, 0, 1);
        truthOutput.setValue(2, 0, 3);

        Matrix optimizedParams = optimizer.optimize(new TargetFunction() {
            @Override
            public BigDecimal fun(Matrix params, Matrix args) {
                BigDecimal x = args.getValue(0, 0);
                BigDecimal a = params.getValue(0, 0);
                BigDecimal b = params.getValue(0, 1);
                BigDecimal res = a.multiply(x).add(b);
                return res;
            }
        }, params, trainInput, truthOutput, true);
        final BigDecimal error = new BigDecimal(0.001);
        Assertions.assertTrue(optimizedParams.getValue(0, 0).subtract(new BigDecimal(2.0)).abs().
                subtract(error).compareTo(new BigDecimal(0.0)) < 0);
        Assertions.assertTrue(optimizedParams.getValue(0, 1).subtract(new BigDecimal(1.0)).abs().
                subtract(error).compareTo(new BigDecimal(0.0)) < 0);
    }

    /**
     * F(X) = a1*x1^2 + a2*x2, 其中a = 1, b = 2.
     * 计算x1 = 4，x2 = 3时F(X)关于a的偏导数
     */
    @Test
    public void testCalcPartialDerivative() {
        Matrix params = new Matrix(1, 2);
        params.setValue(0, 0, 1);
        params.setValue(0, 1, 2);

        Matrix args = new Matrix(1, 2);
        args.setValue(0, 0, 4);
        args.setValue(0, 1, 3);
        BigDecimal partialDerivative = optimizer.calcPartialDerivative(new TargetFunction() {
            @Override
            public BigDecimal fun(Matrix params, Matrix args) {
                BigDecimal x1 = args.getValue(0, 0);
                BigDecimal x2 = args.getValue(0, 1);
                BigDecimal res = params.getValue(0, 0).multiply(x1.multiply(x1)).add(params.getValue(0, 1).multiply(x2));
                return res;
            }
        }, params, 0, args);

        final BigDecimal error = new BigDecimal(0.00001);
        Assertions.assertTrue(partialDerivative.subtract(new BigDecimal(16.0)).abs().subtract(error).
                compareTo(new BigDecimal(0.0)) < 0);
    }


}
